### PR TITLE
Change contract key maintainers to have template argument in scope

### PR DESCRIPTION
**Pull request by martin-drhu-da**
More precisely, change the contract key type class we desugar the surface
syntax to such that the maintainers function has the template argument
in scope instead of the key. The compiler rewrites all references to the
template argument into references to the key and fails whenever that is
not possible.

### CODE REVIEW

**Comment by martin-drhu-da** [comment]
Here's a ticket for the bad error messages: https://github.com/digital-asset/daml/issues/299

**Comment by martin-drhu-da** [comment]
@shayne-fletcher-da Good point regarding `DA.Internal.Desugar`. Added the `TemplateKey` class there.

File: daml-foundations/daml-ghc/src/DA/Daml/GHC/Compiler/UtilLF.hs [hunk 1]
```diff
@@ -160,3 +163,5 @@ ghcTypes = Module
           (ERecCon (TypeConApp (qual (dataTypeCon dataProxy)) [TVar (Tagged "a")]) [])
       }
 
+instance Outputable Expr where
+    ppr = text . renderPretty
```
Comment (neil-da) [inline daml-foundations/daml-ghc/src/DA/Daml/GHC/Compiler/UtilLF.hs]: Seems odd giving a UtilLF module an Ouputable instance. Is this really required?

**Comment by shayne-fletcher-da** [comment]
This PR is associated with issue # https://github.com/DACH-NY/daml/issues/1153

File: daml-foundations/daml-ghc/src/DA/Daml/GHC/Compiler/UtilLF.hs (hunk 1 above)
Comment (martin-drhu-da) [reply daml-foundations/daml-ghc/src/DA/Daml/GHC/Compiler/UtilLF.hs]: It's the `Outputable` instance for `LF.Expr`. The error functions in `Convert.hs` want something `Outputable`. I'll move the instance there.

**Comment by martin-drhu-da** [comment]
I left the tests out because we would need to write them in the desugared syntax now and ideally rewrite them later using syntactic sugar. I've opened an issue to make sure the tests don't fall through the cracks: https://github.com/digital-asset/daml/issues/306.

**Pull request merged**